### DISCUSSION TITLE
Is there an easier way in Python to get the aspect ratio of the current characters on the screen?

### DISCUSSION THREAD

**Discussion by ClaireCJS** [microsoft/terminal#18144] [Q&A]
This seems like an excessive amount of code to determine that my current screen font characters are 1.44X higher than wide (1.44:1 aspect ratio).

is there an easier way to do this?

[Python required, and the Pillow package]

[https://github.com/ClaireCJS/clairecjs_bat/blob/main/BAT-and-UTIL-files/show-screen-font-information-and-character-aspect-ratio.py](https://github.com/ClaireCJS/clairecjs_bat/blob/main/BAT-and-UTIL-files/show-screen-font-information-and-character-aspect-ratio.py)

```
Pillow Version: 10.3.0

--- Default Profile Details ---
altGrAliasing: False
antialiasingMode: grayscale
bellStyle: ['audible', 'window', 'taskbar']
closeOnExit: automatic
colorScheme: Campbell
commandline: c:\tcmd\tcc.exe
cursorShape: filledBox
elevate: True
font: {'face': 'Cascadia Code', 'size': 12}
guid: {dd3810e8-e081-487d-a281-8b3a8c934f95}
hidden: False
historySize: 9001
icon: ms-appx:///ProfileIcons/{61c54bbd-c2c6-5271-96e7-009a87ff44bf}.png
name: TCC
padding: 8, 8, 8, 8
snapOnInput: True
startingDirectory: %USERPROFILE%
useAcrylic: False
-------------------------------

fontFace not found directly. Retrieved 'face' from 'font' dictionary.
fontSize not found directly. Retrieved 'size' from 'font' dictionary.
Font Face: Cascadia Code
Font Size: 12
Font File: C:\Windows\Fonts\CascadiaCode.ttf

--- Measuring Characters ---
Character '|': Width = 7 px, Height = 13 px
Character '⎯': Width = 9 px, Height = 5 px
-----------------------------

Maximum Character Width: 9 pixels
Maximum Character Height: 13 pixels

Height-to-Width Ratio (Scaled): 1.44


DPI Scaling Factor: 1.00

Scaled Maximum Character Width: 9.00 pixels
Scaled Maximum Character Height: 13.00 pixels
```

**Comment by lhecker** [comment]
We currently don't expose the cell size in pixels, although it could be made to partially work in the future by supporting `GetCurrentConsoleFont` in Windows Terminal in the future.
What's the reason you need this information?

**Reply by ClaireCJS** [reply]
> We currently don't expose the cell size in pixels, although it could be made to partially work in the future by supporting `GetCurrentConsoleFont` in Windows Terminal in the future. What's the reason you need this information?

You're going to laugh, but it's to align text into a truly-perfect square to get as much text as possible into a portion of screen real-estate.  

I at least determined that the coefficient for my specific situation and font is 1.4, which will allow me to consider a row to be "1.4 character-widths" high for purposes of aspect ratio calculation and arranging things into a pixel-perfect square for maximizing screen real-estate.

But of course #NotAllFonts are the same ...

**Reply by lhecker** [reply]
> You're going to laugh, but it's to align text into a truly-perfect square to get as much text as possible into a portion of screen real-estate.

Huh... weird. I'm not sure if I'm having a dejavu, but did you post this question here before or perhaps in a Discord? I could've sworn I read this same question a while ago. If it wasn't you: You're not alone. 😅

> I at least determined that the coefficient for my specific situation and font is 1.4

Generally speaking, you can assume that fonts have roughly an aspect ratio of 2:1. It seems you measured the actual size of the glyph though which is slightly different from the cell height/width.

**Reply by ClaireCJS** [reply]
> Huh... weird. I'm not sure if I'm having a dejavu, but did you post this question here before or perhaps in a Discord? I could've sworn I read this same question a while ago. If it wasn't you: You're not alone. 😅

I didn't even know there was a disocrd!

No, this was a new need arising from streamlining my music lyric display transcription process.  I got sick of scrolling up to review input lyrics/output subtitles, so I wrote a little 'newspaper' script to display it in columns like a newspaper.

But of course I couldn't stop there. I like it better if the shape is square, assuming doing that doesn't cause it to take up more rows than are available (in which case, I don't do that)

![image](https://github.com/user-attachments/assets/1f0494b0-6a0e-4bcc-b598-68dc36ba7691)

At first i just checked if ```rows == columns``` but quickly realized the end result was absolutely not square because characters aren't 1:1.


> 
> > I at least determined that the coefficient for my specific situation and font is 1.4
> 
> Generally speaking, you can assume that fonts have roughly an aspect ratio of 2:1. It seems you measured the actual size of the glyph though which is slightly different from the cell height/width.

Arrrrrrrrrrrrrgh that means I failed.

Here's the thing though — I specifically measured the horizontal-connecting-line (NOT hyphen / endash / or emdash) and vertical-connecting-line character (NOT pipe). 

Those characters 100% touch each other when rendered next to each other.... 

so I'm just curious. How are they smaller than the cell size, if they span the whole cell?  Is this ligature rendering voodoo?

**Reply by lhecker** [reply]
They could also be larger than the cell size. It depends on how Pillow renders text (could be GDI for instance). The right way to do it (if anything) is to use the advance width of the "M" character as the cell width, and the line height of the font as the cell height. The line-height is the sum of ascent, descent and line-gap. Pillow does not seem to expose the line-gap, so the next best thing should be something like this:
```py
width, _ = font.getsize("M")
ascent, descent = font.getmetrics()
height = ascent + descent
```

**Reply by ClaireCJS** [reply]
> ```python
> width, _ = font.getsize("M")
> ascent, descent = font.getmetrics()
> height = ascent + descent
> ```

Seems like a great idea, but I just could not get it working. Kept getting ```AttributeError: 'FreeTypeFont' object has no attribute 'getsize'``` ... I think the proper object type is never actually instantiated in this code, and it's actually way above my head. I'm just trying to wield it, i don't think i can understand it 😂
